### PR TITLE
feat(core): make `StackSynthesizer` easier to subclass

### PR DESCRIPTION
**Pull request by rix0rrr**
`StackSynthesizer`s used to make heavy use of private APIs, which make it impossible for people outside the CDK core framework to implement their own Stack Synthesizers and modify the way assets are referenced. This question comes up more and more, so this PR opens up the facilities for doing that.

The changes are as follows:

- Implement `bind()` in `StackSynthesizer`, and add a getter called `boundStack` which will return the bound stack.
- Change `synthesizeStackTemplate -> synthesizeTemplate`, `emitStackArtifact -> emitArtifact`, both of which will implicitly operate on the bound stack.
- Add `addBootstrapVersionRule` so that any subclass can choose to use this or not.
- Expose `AssetManifestBuilder`, which stack synthesizers can use to build an asset manifest. Refactor the class to do less at the same time.
- Expose `cloudFormationLocationFrom...Asset` functions to translate a published asset location into a CloudFormation location.
- Document the contract of each method better.

The end result is that it will be easier to mix and match the different behaviors of the existing synthesizers together into new custom behavior.

Two new unit test files have been added to show how it is possible to write ONE intruction to the asset manifest (to upload assets to a given location) while returning a completely different instruction to the template (referencing the asset bucket using a CloudFormation parameter).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by Naumel** [comment]
@Mergifyio update

File: packages/@aws-cdk/aws-ecr-assets/test/custom-synthesis.test.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * CloudFormation template -- without reaching into the library internals
+ */
+
+import * as path from 'path';
+import { Template } from '@aws-cdk/assertions';
+import { StackSynthesizer, FileAssetSource, FileAssetLocation, DockerImageAssetSource, DockerImageAssetLocation, ISynthesisSession, App, Stack, AssetManifestBuilder, CfnParameter, CfnResource } from '@aws-cdk/core';
+import { AssetManifestArtifact } from '@aws-cdk/cx-api';
+import { DockerImageAsset } from '../lib';
+
+// const DEMO_IMAGE_ASSET_HASH = '0a3355be12051c9984bf2b0b2bba4e6ea535968e5b6e7396449701732fe5ed14';
```
Comment (Naumel) [inline packages/@aws-cdk/aws-ecr-assets/test/custom-synthesis.test.ts:13]: Left over or should have more context to be considered part of docs?

File: packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+      executable: asset.executable,
+      directory: asset.directoryName,
+      dockerBuildArgs: asset.dockerBuildArgs,
+      dockerBuildTarget: asset.dockerBuildTarget,
+      dockerFile: asset.dockerFile,
+      networkMode: asset.networkMode,
+      platform: asset.platform,
+    }, {
+      repositoryName: target.repositoryName,
+      imageTag,
```
Comment (Naumel) [inline packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts:74]: [Q] This is equivalent to `imageTag: imageTag,`?

File: packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+      };
+    }
+    this.files[sourceHash].destinations[this.manifestEnvName(stack)] = dest;
+    return dest;
+  }
+
+  /**
+   * Add a docker asset source and destination to the manifest
+   *
+   * sourceHash should be unique for every source.
```
Comment (Naumel) [inline packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts:100]: Where does this come in for the code below?
There is no check for 
```
if (this.dockerImages[sourceHash]) {
    if (!this.dockerImages[sourceHash].source === source) {
(...)
```

**Review by Naumel — APPROVED** [review]
Looks good, added the `do-not-merge` label in case you'd like another approval.

File: packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts:74]: Yes!

File: packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts (hunk 3 above)
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/stack-synthesizers/asset-manifest-builder.ts:100]: That's a prerequisite we don't check for.

**Pull request merged**